Fix ValueError with tf.keras.metrics.Recall and float64 keras backend

yongtang opened the pull request:
This PR fixes the issue raised in #36790 where tf.keras.metrics.Recall
causes ValueError when the backend of the keras is float64:

This PR cast the value to the dtype of var as var.assign_add
is being called.

This PR fixes #36790.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix ValueError with tf.keras.metrics.Recall and float64 keras backend

This PR fixes the issue raised in 36790 where tf.keras.metrics.Recall
causes ValueError when the backend of the keras is float64:

This PR cast the value to the dtype of var as var.assign_add
is being called.

This PR fixes 36790.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for tf.keras.metrics.Recall() and float64 keras backend.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

pavithrasv commented on tensorflow/python/keras/utils/metrics_utils.py:
```diff
@@ -422,9 +422,9 @@ def update_confusion_matrix_variables(variables_to_update,
   def weighted_assign_add(label, pred, weights, var):
     label_and_pred = math_ops.cast(
-        math_ops.logical_and(label, pred), dtype=dtypes.float32)
+        math_ops.logical_and(label, pred), dtype=var.dtype)
```
Could we change other places float32 is used in this function as well?

pavithrasv reviewed (commented):
Thank you for the PR!

yongtang committed:
Update update_confusion_matrix_variables to alwasy cast to variables_to_update dtype (vs. explicit float32)

This commits updates the function update_confusion_matrix_variables
to alwasy cast to dtype based on  variables_to_update (previously
the values are casted to float32 explicitly and that cuases issues
when keras' backend use non-float32).

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @pavithrasv for the review. The PR has been updated and now all float32 cast in the function have been changed to use the dtype of the variable. Please take a look and let me know if there are any issues.

pavithrasv reviewed (approved):
Thank you!
